### DISCUSSION TITLE
Topological Enthalpy Model for Tautomers

### DISCUSSION THREAD

**Discussion by pwging13** [rdkit/rdkit#7887] [Ideas]
So I've been using the tautomer canonicalization within rdkit, and the scoring function isn't perfect, "canonical" doesn't necessarily mean most energetically favorable.  "Most energetically favorable" is also context dependent.

However, here's a molecule from chembl that I think highlights how physically unreasonable (at least according to my intuition) the scoring function can be.

![image](https://github.com/user-attachments/assets/04c0eb38-7fab-4a6b-887c-c2f7340d3216)

While that methyl proton very likely as a reduced pKa value, I don't think that exocyclic double bond would be appreciably observed.

Instead of relying solely on the scoring function, how about a simple, scalable linear model that partitions the enthalpy from a gas phase electronic structure calculation?  Then, using only the topology of a molecule, we could estimate the total enthalpy of a tautomer?

I've built such a model over about 500k tautomers from ChEMBL plus another 200k random sampled molecules from ChEMBL.  From each SMILES, I sampled conformers and selected the lowest energy conformation according to MMFF94s within rdkit.  Then, I ran that conformation through xtb using GFN2-xTB to get the enthalpy.  GFN2 is obviously a semiempirical method, but it quickly allows for this large sampling of chemical/tautomer space.

I then segmented all the bond types, according to atom types, hybridization, bond orders, conjugation, aromaticity, and whether the bond is part of a ring.
![image](https://github.com/user-attachments/assets/4bc36d4e-a270-4a90-abb6-f6e31e8d03af)

According to these ~600 bond types, each molecule is featurized and a linear model was trained using ridge regression.  On a 90/10 split, here's the test set performance.  The MAE is about 9 kcal/mol, and the MAPE is 0.02%.  I think this is pretty good considering the simplicity of the approach.
![Test_Set_Performance](https://github.com/user-attachments/assets/bed91caa-2aa0-4598-a5db-6f1484a08ec6)

This model then selects what I think is a better tautomer for the molecule above
![image](https://github.com/user-attachments/assets/fffb4c0a-ed34-4425-9da5-c7e52fde8616)

I then looked at the correlation to better enthalpies from the QM9 dataset, and I think the correlation is pretty reasonable given the model only considers topology and was built on a MUCH cheaper method.
![image](https://github.com/user-attachments/assets/478856b9-99cb-4757-9983-8431b0b6e2f6)

Interested in feedback on this.  A more complicated GNN might reduce the MAPE further, but at appreciable compute costs.  This just builds on the old school idea of adding up bond enthalpies.

**Comment by DavidACosgrove** [comment]
Canonical doesn’t mean anything more than that if you give a compound in one of its possible tautomers you will always get the same tautomer out. It’s to address the problem of comparing two sets of compounds to see which are in common, for example.  The rules attempt to find a sensible tautomer but it’s crude. Apart from anything else the “correct” tautomer is a function for solvent, temperature, pH etc.

**Reply by pwging13** [reply]
@DavidACosgrove I acknowledge your points.  I'm asking if canonicalization, which depends on the scoring function, can be made more physically or energetically sensible and less crude.  I'm proposing the above approach as a scalable, topologically driven way to accomplish that.

**Comment by bp-kelley** [comment]
I agree with David, but I would love an approach that generates a reasonable tautomer at pH 7 and would be happy to support one even if we add it as a contribution initially.

**Comment by greglandrum** [comment]
@pwging13 the work looks interesting! It's great that you have built a simple linear model instead of something requiring a GNN (or even a random forest). It does seem like it would be more applicable if the calculations were being done in implicit solvent (or did I miss something?); it makes the calculations more expensive, but can make a real difference for this kind of thing. The logical next step could be to do calculations in a number of different solvents to allow prediction of a "most stable" tautomer in different environments. 

If we were to add something like this to the RDKit, I would suggest leaving the current canonical tautomer code as-is (that's intended to solve a cheminformatics problem) and add this (or future related methods) as alternative tautomer scoring functions. That would give the people who need the canonical tautomer for de-duplication stability but would still allow us to have one or more approaches which attempt to provide the "most stable" tautomer for a particular set of conditions (and which could be easily updated as models/understanding improve).

**Comment by pwging13** [comment]
@greglandrum implicit solvent could help.  I ran this in gas phase as a first pass, but solvation could be added.

Thinking about how to do that, I reran a smaller set of molecules and recorded the free energy and entropy in addition to the enthalpy. I'll regress against the free energy instead of the enthalpy.  I expected this to be worse from the entropy contribution, but this still worked pretty well.
![Test_Set_g298_Performance](https://github.com/user-attachments/assets/c7e37c24-f87f-4f56-a4f0-24ee4ca56c4e)

I just need to perform the single point calculations in water, DMSO, chloroform, etc to get each molecule's Gsoln and train the respective models.  Those calculations will take a little bit of time, but not too long as I already have the optimized gas phase geometries.  Will report back.

**Reply by pwging13** [reply]
Meeting expectations, predicting the entropy is worse
![Test_Set_s298_Performance](https://github.com/user-attachments/assets/00f80ac5-13ae-4e59-93ff-950211305593)

**Reply by bp-kelley** [reply]
What's the RMSE there?  It looks << 1kcal/mol which I think is pretty good.

**Reply by bp-kelley** [reply]
I have a set of hard to predict tautomers I can run it on as a simple benchmark (sadly, I can't report some of these) but would be quite happy to.

**Reply by pwging13** [reply]
@bp-kelley how many can you share here as SMILES?  It would be interesting if you deliberately share a known wrong tautomer, I can run them, and report back the energetically preferred tautomer.

**Reply by pwging13** [reply]
> What's the RMSE there? It looks << 1kcal/mol which I think is pretty good.

I'm expanding the dataset and adding implicit solvation effects, so the below are for gas phase and not from the entire dataset

The MAE for enthalpy: 10.3 kcal/mol
The MAE for free energy: 10.2 kcal/mol
The MAE for entropy: 0.6 kcal/mol -> this is of course much large expressed as a relative value, hence the MAPE of over 1%

**Reply by bp-kelley** [reply]
@pwging13 at this point the answer is I can release "none" but I can maybe get a few together.

**Reply by pwging13** [reply]
@bp-kelley I understand how that goes.

**Reply by pwging13** [reply]
about 90% of the 1.5 million molecules are done, and I've taken a first pass as training a model on the free energy and enthalpy in water.

Note: the geometry optimization and vibrational analysis was performed in implicit solvent.  I acknowledge there are varying schools of thought on this given the partition function is meant for gas phase.  That being said, for the sake of this model, I'll assume errors arising from this assumption will cancel given we're interested in comparative analysis.

For enthalpy in water:
![Test_Set_h_water_Performance](https://github.com/user-attachments/assets/df10d255-864b-4f2d-aea7-7b3683f73c47)

For free energy in water: 
![Test_Set_g_water_Performance](https://github.com/user-attachments/assets/d139f04d-a2c1-4f51-8c59-ede64c17e609)

Enthalpy in chloroform
![Test_Set_h_chcl3_Performance](https://github.com/user-attachments/assets/ca0246f7-bd67-4797-9b54-1c1bb810d6c5)

Free energy in chloroform:
![Test_Set_g_chcl3_Performance](https://github.com/user-attachments/assets/b9cc8864-19b9-484a-ba3b-d34c713d260d)

Enthalpy in octanol:
![Test_Set_h_oct_Performance](https://github.com/user-attachments/assets/7b7e24d6-8869-41d4-a2da-de6708207533)

Free energy in octanol:
![Test_Set_g_oct_Performance](https://github.com/user-attachments/assets/079068d2-ff57-46fa-a5e2-e40b92281c38)

I chose octanol to perhaps sanity check this against LogP values. Haven't started that yet.

**Comment by pwging13** [comment]
@greglandrum performance looks good on gas phase free energies corrected by Gsolv from single point calculations using ALPB within xtb.  If we optimize the structures and perform vibrational analysis in implicit solvent, the performance is basically the same.  Adding solvents with corrections to gas phase free energies is obviously much cheaper.

I'll probably write this up, though I doubt it will land anywhere overly impactful.  If this is useful to the community, I'm happy to share code. ahead of that.

Gas Phase:
![Test_Set_g_gas_Performance](https://github.com/user-attachments/assets/082d2e27-7e59-45d0-a14e-2a3624db9be5)

Water:
![Test_Set_ggas_plus_gsolvwater_Performance](https://github.com/user-attachments/assets/e140a911-012e-47cc-9165-fab7ca5bdafd)

Chloroform:
![Test_Set_ggas_plus_gsolvchcl3_Performance](https://github.com/user-attachments/assets/17a1cce7-f823-48c5-98de-4ea08c5a75e3)

DMSO:
![Test_Set_ggas_plus_gsolvdmso_Performance](https://github.com/user-attachments/assets/87346706-0ff2-43c0-ac10-6626946f6043)

Hexane:
![Test_Set_ggas_plus_gsolvhex_Performance](https://github.com/user-attachments/assets/bc544500-c1c7-4fdf-9cbc-2c0695655d13)

Methanol:
![Test_Set_ggas_plus_gsolvmethanol_Performance](https://github.com/user-attachments/assets/36dde76b-8fcf-4581-9597-d36dd6547562)

Toluene:
![Test_Set_ggas_plus_gsolvtol_Performance](https://github.com/user-attachments/assets/84c913c8-f19c-49fc-9459-9275eee6ba3c)

Octanol:
![Test_Set_ggas_plus_gsolvoct_Performance](https://github.com/user-attachments/assets/6f9bda72-4344-437c-ba2d-b7ba3a1b85b1)
